### PR TITLE
Real time updating pull-information on review site

### PR DESCRIPTION
**Pull request by MichaelMayorov**
New:
1. Suport for `pull_request` hook
2. `/upload_pull/` for secret URL generation and uploading pull requests from github
3. Secret URL will be emailed to all admins after each generation
4. Logging
5. Manual about setting up

I didn't deleted old pull updating mechanism, because new functional should be tested by other community members before adding to master.

### CODE REVIEW

**Comment by certik** [comment]
Let's test this before we merge it. I've uploaded your branch here:

http://repos.sympy.org/

@MichaelMayorov, can you send me your google id so that I can add you to the app engine test application above? That way you can manage it as well.

**Comment by certik** [comment]
So the admin page works and I've added the generated URL to hooks. I then did:

```

$ curl -u certik -d '{ "events": [ "pull_request" ] }' https://api.github.com/repos/sympy/sympy/hooks
Enter host password for user 'certik':
{
  "message": "Validation Failed",
  "errors": [
    {
      "value": null,
      "field": "name",
      "resource": "Hook",
      "code": "invalid"
    }
  ]
}
```

Any idea why it fails?

**Comment by certik** [comment]
I reviewed the code and it looks good in general. However I would like to test it first before we merge it and upload to our official page, so we need to resolve the hook creation at github.

Also I have some minor grammar changes, I'll post those once things are working.

File: web/templates/upload_url.html [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        <p>
+        {% if upload_url %}
+            <a href="/upload_pull/{{ upload_url }}">{{ upload_url }}</a>
+        {% else %}
+            There is no upload URL yet, press "Generate" to create new
+        {% endif %}
+        </p>
+        <b>NOTE:</b>
+        <p>
+        If you will press button below, then review site will generates new URL
```
Comment (asmeurer) [inline web/templates/upload_url.html]: will generate a new URL

File: web/templates/upload_url.html [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        for uploading pull request information and will send email with new URL
```
Comment (asmeurer) [inline web/templates/upload_url.html]: send an email

File: web/templates/upload_url.html [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        to all admins
```
Comment (asmeurer) [inline web/templates/upload_url.html]: all admins. (add period/full stop)

File: web/templates/upload_url.html [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                        <td>{{ row.created_at|date:"r" }}</td>
+                    </tr>
+                {% endfor %}
+            </tbody>
+        </table>
+
+    {% else %}
+        Your account <b>{{ user.nickname }}</b> has no rights to access this page
+        <br />
+        May be <a href="{{ logout_url }}">logout</a> and try again?
```
Comment (asmeurer) [inline web/templates/upload_url.html]: Maybe is one word.

File: README.rst [hunk 5]
```diff
@@ -95,3 +95,31 @@ of the commit you want.
...
+but that will update information about pulls in real time.
+
+SymPy Bot web-interface(which located in under web/) supports integration with
+github via mechanism called hooks http://developer.github.com/v3/repos/hooks/
+
+To use that feature you need to follow these steps:
+
+1. Go to ``http://example.com/upload_pull``, sign in as administrator and press
+   ``generate`` button. After that, all admins will recieve notification with
+   secret URL(you can see a log of all generations in table on that page)
```
Comment (asmeurer) [inline README.rst]: Space before (

Comment (asmeurer) [inline README.rst]: And elsewhere

**Comment by MichaelMayorov** [comment]
> Any idea why it fails?

@certik Yes, you need to speciffy hook id.

So, full string will looks like:

> curl -u MichaelMayorov -d '{ "add_events": [ "pull_request" ] }' https://api.github.com/repos/MichaelMayorov/test-hooks/hooks/316906

That's my mistkake, because I copy-pasted wrong command from my bash history :(

**Comment by certik** [comment]
But how do I know the hook id?

**Comment by MichaelMayorov** [comment]
@certik You could try:

> curl -u username https://api.github.com/repos/username/repo/hooks

Will list all hooks. You'll need `id` field

**Comment by certik** [comment]
It works! Can you please push in a patch with the following instructions (in the README):

```
Find the hook that you want to modify by::

    curl -u username https://api.github.com/repos/username/repo/hooks

the ``id`` field gives the hook ID, copy and paste the path in the "url" field into the command::

    curl -u username -d '{ "add_events": [ "pull_request" ] }' https://api.github.com/repos/username/repo/hooks/ID

You will see that the "events" part::

    "events": [
      "push"
    ],

changed to::

  "events": [
    "push",
    "pull_request"
  ],

```

This should make it clear how it works.

**Comment by certik** [comment]
I created a test pull request: https://github.com/sympy/sympy/pull/1403

We also need to test, that if you add comments to the pull request, that we get notified by github.

**Comment by certik** [comment]
Hm, I don't see it there. Any ideas? @MichaelMayorov please send me your gmail id, so that I can add you to the appengine app and you can debug it yourself.

**Comment by MichaelMayorov** [comment]
@certik my gmail account is marchael@kb.csu.ru

But I think, that it'd be easier to debug if you add hook to 

> http://shell.marchael.ru/upload_pull/38137214a33e38c2804d9260e86fb92f81826ae7

It's my laptop and I can see dev server logs.

**Comment by MichaelMayorov** [comment]
For now I see,

> Traceback (most recent call last):
>   File "/home/marchael/Downloads/google_appengine_/google/appengine/ext/webapp/_webapp25.py", line 703, in __call__
>     handler.post(*groups)
>   File "/home/marchael/repos/sympy-bot-my/sympy-bot/web/app/main.py", line 271, in post
>     num = payload["number"]
> KeyError: 'number'

**Comment by MichaelMayorov** [comment]
@certik Is that pull request or commit? Usually pull requests have "num" field("num" is number of pull)

**Comment by certik** [comment]
I don't know what github sends. I just setup the regular web hook and enabled the "pull_request" event. My bet however is that if someone pushes in a commit, it will be sent as a commit?

**Comment by certik** [comment]
So here is the log that I can see on the appengine at 2012-07-03 20:47:57.418:

```
The API call datastore_v3.Put() required more quota than is available.
Traceback (most recent call last):
  File "/base/python_runtime/python_lib/versions/1/google/appengine/ext/webapp/_webapp25.py", line 703, in __call__
    handler.post(*groups)
  File "/base/data/home/apps/sympy2/4.360064092256910398/app/main.py", line 300, in post
    p.put()
  File "/base/python_runtime/python_lib/versions/1/google/appengine/ext/db/__init__.py", line 1074, in put
    return datastore.Put(self._entity, **kwargs)
  File "/base/python_runtime/python_lib/versions/1/google/appengine/api/datastore.py", line 579, in Put
    return PutAsync(entities, **kwargs).get_result()
  File "/base/python_runtime/python_lib/versions/1/google/appengine/api/apiproxy_stub_map.py", line 604, in get_result
    return self.__get_result_hook(self)
  File "/base/python_runtime/python_lib/versions/1/google/appengine/datastore/datastore_rpc.py", line 1579, in __put_hook
    self.check_rpc_success(rpc)
  File "/base/python_runtime/python_lib/versions/1/google/appengine/datastore/datastore_rpc.py", line 1214, in check_rpc_success
    rpc.check_success()
  File "/base/python_runtime/python_lib/versions/1/google/appengine/api/apiproxy_stub_map.py", line 570, in check_success
    self.__rpc.CheckSuccess()
  File "/base/python_runtime/python_lib/versions/1/google/appengine/api/apiproxy_rpc.py", line 133, in CheckSuccess
    raise self.exception
OverQuotaError: The API call datastore_v3.Put() required more quota than is available.
```

That just means that it is currently overquota. I reused some older application that I didn't use, and it was simply running overquota, so it's not our fault. We just need to wait probably a day, until it settles.

However, right after the above, I can also see the following failure at 2012-07-03 20:53:40.105:

```
'number'
Traceback (most recent call last):
  File "/base/python_runtime/python_lib/versions/1/google/appengine/ext/webapp/_webapp25.py", line 703, in __call__
    handler.post(*groups)
  File "/base/data/home/apps/sympy2/4.360064092256910398/app/main.py", line 271, in post
    num = payload["number"]
KeyError: 'number'
```

So we need to figure out what is going on.

**Comment by MichaelMayorov** [comment]
> {
>   "config": {
>     "url": "http://shell.marchael.ru/upload_pull/38137214a33e38c2804d9260e86fb92f81826ae7",
>     "content_type": "form",
>     "insecure_ssl": "1"
>   },
>   "last_response": {
>     "status": "ok",
>     "message": "OK",
>     "code": 200
>   },
>   "active": true,
>   "updated_at": "2012-07-04T03:56:45Z",
>   "events": [
>     "push",
>     "pull_request"
>   ],
>   "name": "web",
>   "id": 123,
>   "url": url,
>   "created_at": "2012-07-04T03:53:24Z"
> }

Here:

>   "events": [
>     "push",
>     "pull_request"
>   ],

AFAIK, web application now and before didn't process `push` events. I guess we should leave only `pull_request` for now.

That could be done with

>   curl -u username -d '{ "events": [ "pull_request" ] }' https://api.github.com/repos/username/repo/hooks/ID

**Comment by certik** [comment]
So here is my forensic analysis: The first failure failed at the line 300, which means that your code went all through, except the database write, which run over quota (not our fault). 

The second failure failed much earlier. This must have something to do with a different thing that github sent us. We are currently listening on two events: "push" and "pull_request". So we need to remove the "push" event --- as you just figured out!

**Comment by certik** [comment]
Done --- I've just left the "pull_request" events for both your and appengine repos. Let me add a new commit to my test pull request to see what happens.

**Comment by certik** [comment]
Here is the log after my commit:

https://gist.github.com/3045229

I think it works!

**Comment by MichaelMayorov** [comment]
Yes, I see update in "Non-mergeable Pullrequests"

**Comment by MichaelMayorov** [comment]
It's also available here http://shell.marchael.ru/pullrequest/1403/

**Comment by certik** [comment]
Cool. So the code clearly works. Thanks!

Would you mind fixing the little issues that Aaron reported? Also, if you could please document in the README (per my comment above) how to setup the hook properly at github --- i.e. that you need to use the command of the form:

```
curl -u username -d '{ "events": [ "pull_request" ] }' https://api.github.com/repos/username/repo/hooks/ID
```

After that, let's merge it.

**Comment by MichaelMayorov** [comment]
All right, I will fix it. But I think we still need more proper testing with real pull requests

**Comment by MichaelMayorov** [comment]
I found that "Author name": "None" ugly, and seems for me, the only way to receive that - poll github for user account detailts.

May be replace "Author name" to "Author login" and use appropriate field from pull request?

**Comment by certik** [comment]
Sure, let's test it more. In that case, push in the following patch:

```

diff --git a/web/app.yaml b/web/app.yaml
index e0f3eb3..7bb5c5b 100644
--- a/web/app.yaml
+++ b/web/app.yaml
@@ -1,4 +1,4 @@
-application: sympy3
+application: sympy2
 version: 4
 runtime: python
 api_version: 1
diff --git a/web/cron.yaml b/web/cron.yaml
deleted file mode 100644
index a6da74c..0000000
--- a/web/cron.yaml
+++ /dev/null

@@ -1,7 +0,0 @@
-cron:
-- description: update from GitHub (quick)
-  url: /quickupdate
-  schedule: every 12 hours
-- description: update from GitHub (full)
-  url: /update
-  schedule: every day 00:00
```

It removes the cron jobs, that way the old mechanism can stay in place, as it will never be used. And it changes which appengine apps should be updated --- the sympy2 is the repos.sympy.org one.

Then after you commit into this pull request, just push in the changes by:

```
apcfg update .
```

and enter your google ID and password.

We will test it at repos.sympy.org for a while. If it works, I'll give you an access to the main app at reviews.sympy.org and after this is merged, you can push in there a new version.

Also actually a good idea is to always increase the "version" in app.yaml, before you push it in.

**Comment by MichaelMayorov** [comment]
> Then after you commit into this pull request, just push in the changes by:

I've done it

> Completed update of app: sympy2, version: 4

Is that correct?

**Comment by MichaelMayorov** [comment]
btw, I found that cron jobs weren't removed

**Comment by certik** [comment]
The cron jobs should get removed by removing the cron.yaml before uploading.

**Comment by MichaelMayorov** [comment]
I removed and reupload again, but cron jobs still exists. Maybe that's happenning because we don't change version?

**Comment by certik** [comment]
Hm, I don't know. I changed the version to 5, reuploaded, went online, changed to version 5, deleted version 4. Still the cron jobs are there... Hm. Maybe we need to add an empty cron.yaml file? Time to lookup the docs:

https://developers.google.com/appengine/docs/python/config/cron#Uploading_Cron_Jobs

Yes. So I did that and the cron jobs are done! Finally.

As to your other questions about "None": For now I would just keep updating the app fully using the github hooks. Let's use just the login, that's fine for now. So I would just save everything that we can from the github payload, and then when showing the pull request, if name is None, then just show the login.

Then, if you want, when github sends us the payload, schedule a task for updating the user if the name is None. Typically, the user will already be in the database, so no update is needed.

Last problem to solve is when people update their user data at github, whether there is any way to get our app notified, or whether once in a while, a full refresh needs to be done.

For all these reasons, I would keep the old update mechanisms in place, so that we can modify them later.

**Comment by MichaelMayorov** [comment]
hm, I don't think that names is a problem which we should worried now.

I suggest to replace names to logins with refers on their github repos, like https://github.com/user/repo

That way have one, but usefull advantage - we don't need to poll github for user names, instead of that we could just add html href tag to github account and that's all.

**Comment by certik** [comment]
That's definitely fine with me.

**Comment by MichaelMayorov** [comment]
I alredy upload current code to http://repos.sympy.org

**Comment by MichaelMayorov** [comment]
Well, so far as I seen we have a one problem with these hooks. Rather not with hooks, but with export existing pull requests.

So, I suggest to use polling functional for initial export pull requests from github to review site. To do this, we could  add submit button "Populate" to `/upload_pull`

**Comment by MichaelMayorov** [comment]
OK, I added new functional for initial uploading all pulls to review site, needs testing on clean database.

**Comment by certik** [comment]
It looks good. If you have time, please fix the Aaron comments above (some spaces are missing).

In order to test this one, feel free to delete the whole database of the `sympy2` project, see here: http://stackoverflow.com/questions/1062540/how-to-delete-all-datastore-in-google-app-engine

**Comment by certik** [comment]
I think you will have to write a simple script to erase the whole database. If so, post it to gist, so that we can reuse it the next time.

**Comment by MichaelMayorov** [comment]
@certik First I tried delete all entries on my own test base sympy-bot-test.appspot.com, but that has led me to exhausted quota limit :(

**Comment by MichaelMayorov** [comment]
btw, seems tables could be cleared via "Datastore Admin". So now, I try to enable and play with it.

**Comment by MichaelMayorov** [comment]
Ok, done it. But that's not all pull requests. Seems for me, github returned only recent pull requests.

**Comment by MichaelMayorov** [comment]
@certik I found strange thing in github hooks. If you'll see on recent pull requests on http://repos.sympy.org and http://reviews.sympy.org then you should saw that, e.g. pull http://repos.sympy.org/pullrequest/1407/ marked as "Non-mergeable" on repos and "Mergeable" on reviews.

I tried to find out why this happens and looked in logs, and found that last record about that pull contained `"mergeable" : "None"`. But when I did 

> curl https://api.github.com/repos/sympy/sympy/pulls/1408

It returned `"meargeable" : "True"`

**Comment by certik** [comment]
Right. I think that mergeable "None" means that github didn't check (yet) if it is mergeable. Later it decides either yes or no, but in repos.sympy.org we don't get this update. I would ignore this problem for now.

One thing that we can do is to also update the request whenever somebody comments in it --- that should be the `issue_comment` hook. That way we can update the metadata in the pull request, and also allow (for later use) to have code in place to react if the user wants to instruct sympy-bot using comments like "sympy-bot test" or something. What do you think?

In any case, I think that we can merge this pull request and set this up in our official app. Do you agree?

**Comment by MichaelMayorov** [comment]
@certik 

> that should be the issue_comment hook. That way we can update the metadata in the pull request

hm, how exactly this hook will help us? Payload doesn't content anything usefull for current state of web app. Maybe later we could use this, but for now it doesn't have much sense.

Another case for that - poll github for "mergable" status after `issue_comment` triggered.

**Comment by MichaelMayorov** [comment]
> In any case, I think that we can merge this pull request and set this up in our official app. Do you agree?

Of course I am :)
But I think that this pull request should be closed only when web app will works correct.

**Comment by certik** [comment]
I agree with your comment about `issue_comment`, let's do it later. Except --- are you sure the `issue_comment` payload doesn't contain the "mergeable" field? Because if it does, then we can update mergeable field of the PR that way.

The app seems to be working fine. Is there anything that you are aware of that isn't working?

(For example I closed a pull request and the app registered it automatically and moved the PR into closed ones.)

I think the only thing is the mergeable issue. I noticed that mergeable = None show up under "non-mergeable PRs". @MichaelMayorov, if you agree, can you push in the following patch:

```

diff --git a/web/app/main.py b/web/app/main.py
index 6420d3c..6cbe65b 100644
--- a/web/app/main.py
+++ b/web/app/main.py
@@ -72,11 +72,11 @@ class MainPage(RequestHandler):
             last_quick_update = p.last_updated
             last_quick_update_pretty = pretty_date(last_quick_update)
         p_mergeable = PullRequest.all()
-        p_mergeable.filter("mergeable =", True)
+        p_mergeable.filter("mergeable IN", [True, None])
         p_mergeable.filter("state =", "open")
         p_mergeable.order("-created_at")
         p_nonmergeable = PullRequest.all()
-        p_nonmergeable.filter("mergeable IN", [False, None])
+        p_nonmergeable.filter("mergeable =", False)
         p_nonmergeable.filter("state =", "open")
         p_nonmergeable.order("-created_at")
         self.render("index.html", {
```

It should move the "None" PRs into mergeable, which I think is where most such PRs should belong. We can add a simple explanation into the html, that this section also contains mergeable=None PRs.

**Comment by MichaelMayorov** [comment]
@certik 

> are you sure the issue_comment payload doesn't contain the "mergeable" field?

Yes, I tried it on my test repo.

> The app seems to be working fine. Is there anything that you are aware of that isn't working?

Yes, I found that github doesn't return all pull requests(only recent)

>  It should move the "None" PRs into mergeable, which I think is where most such PRs should belong. We can add a simple explanation into the html, that this section also contains mergeable=None PRs.

Well, that's good idea, but maybe create different table for these pulls?

**Comment by asmeurer** [comment]
> Well, that's good idea, but maybe create different table for these pulls?

I agree with that plan.

**Comment by certik** [comment]
Ok, let's create another table for merge None.

> Yes, I found that github doesn't return all pull requests(only recent)

Can you clarify what that means? That github doesn't send us updates on older pull requests when they are updated? Or are you talking about initial population of the database?

**Comment by MichaelMayorov** [comment]
> Can you clarify what that means?

That means that when web app do something like `curl https://api.github.com/repos/sympy/sympy/pulls`, then github response only pulls, which were created 2-3 months ago (just look at http://repos.sympy.org)

> That github doesn't send us updates on older pull requests when they are updated?

post-receive hooks work as for old as well as for new pull requests.

> Or are you talking about initial population of the database?

exactly.

**Comment by certik** [comment]
Try to populate it again --- I think there was some problem with running over quota on the app engine. I am quite sure that github sends all the pull requests, at least it used to.

**Comment by MichaelMayorov** [comment]
well, I cleaned database on http://repos.sympy.org and made initial populating, but as you can see, there is only recent pulls.

**Comment by MichaelMayorov** [comment]
btw, I watched at work of these hooks and found that github always post `"meargeable" : "None"` when pull request is creating or updating, so that means that we'll never know current "mergeabe" state for all incoming pull requests and updates. (I just watching logs for http://repos.sympy.org)

So, I suggest to poll github for pull request "mergeable" state each time when hook `pull_request` triggered.

**Comment by MichaelMayorov** [comment]
I pushed these changes to http://repos.sympy.org 
Now initial populating exports all open pull requests and there are no pull requests with `mergeable: None`

I'm wonder why github always post data with `mergeable: None`. Maybe, it's a bug?

**Comment by asmeurer** [comment]
Probably because it hasn't computed it yet when the pull request is first created. It's also possible that it doesn't compute it at all until someone with push access loads the page.

**Comment by MichaelMayorov** [comment]
@asmeurer perhaps. But the problem is, do we need to poll github each time when hook triggered or not? 

If not, then after some time, most part of pull requests on review site will have `mergeable: None` status

In my last commit web app polls github each time when hook triggered. it's not grace solution, but it works. And I want to ask you or @certik to test my code on http://repos.sympy.org (I alredy pushed new version here)

**Comment by asmeurer** [comment]
The closed pull request page is empty.

**Comment by MichaelMayorov** [comment]
This page is empty because initial populating didn't download information about closed pull requests(but after closing on github pull requests will be appear there).

We could export closed pull requests too, but why?

**Comment by asmeurer** [comment]
I guess if someone wants to find old reviews on an old request. It probably isn't top priority, though.

**Comment by MichaelMayorov** [comment]
Well, that's make sense.

But, populate button should be used only on fresh installs without current pull requests database. In this case, we don't need to export closed pull requests because they don't have any reviews.

From the other hand, sites like http://reviews.sympy.org already have pull requests database and you don't need to make initial export.

**Comment by certik** [comment]
Ok, I can see why the "populate" button only populates a few requests. The `WorkerNG` implementation will not work, because there is about 30s limit per request on the app engine, and that is not enough to process all pull requests by querying github for each of them. So the appengine simply kills the process in the middle. The way to fix is to do the same thing as in the UpdatePage handler, that is, schedule a task for every single pull request.

**Comment by MichaelMayorov** [comment]
@certik hm, do you see current http://repos.sympy.org ? Seems for me, initial populate exported all pull requests.

> The WorkerNG implementation will not work, because there is about 30s limit per request on the app engine

Maybe, but seems true reason of it was behaviour of github.
When someone do something like `curl https://api.github.com/repos/sympy/sympy/pulls` github split large amount of pull requests on chunks or pages (to prevent channel from overloading I guess) and returns only first page. To have access to other page you need to parse one of the HTTP headers.

Just type `curl https://api.github.com/repos/sympy/sympy/pulls`in your terminal and try to find one of the old pull requests. You could also look at web/app/github.py code to see how works github_get_pull_request_all_v3()

**Comment by certik** [comment]
Yes, that's why we have the `get_all_pages()` function in `web/app/github.py`.

I am misunderstanding what doesn't work -- I thought you initially said that the initial "population" of the pull requests doesn't work, but now you seem to be saying that it works. So I am confused.

In any case, if you think that things work, let's merge it. I just want an ok from you, before I merge it.

**Comment by MichaelMayorov** [comment]
> I am misunderstanding what doesn't work -- I thought you initially said that the initial "population" of the pull requests doesn't work, but now you seem to be saying that it works. So I am confused.

@certik Well, before https://github.com/MichaelMayorov/sympy-bot/commit/f3114a6ffeeefc1d6e673bfb69b90bcc9b82047e it doesn't work as expected :)

**Comment by MichaelMayorov** [comment]
I uploaded new changes to http://repos.sympy.org ,seems working.
@asmeurer @certik I guess this code is ready for merging to master. Is anything I can do to improve it?

**Comment by MichaelMayorov** [comment]
Is someone still interesting in this pull?

**Comment by certik** [comment]
@MichaelMayorov, I've been flying to Prague, so I was off the net for 2 days. Let me have a look.

**Comment by certik** [comment]
This looks good. I am merging it and I'll set it up.

**Comment by certik** [comment]
Ok, it should be setup at: http://reviews.sympy.org/, I've updated the hooks at github. Can you check somehow that it actually works? I've added you to developers of the webapp, so that you can see the logs.

**Comment by MichaelMayorov** [comment]
> Can you check somehow that it actually works?

@certik Yes, all stuff works correct. I think that next step we should make clean up in web app, that do you think?

**Comment by certik** [comment]
@MichaelMayorov, definitely. Feel free to send more PRs.

**Comment by asmeurer** [comment]
@certik assumedly that was #118.

**Pull request merged**